app/controllers/api/v0/events_controller.rb

```diff
@@ -16,23 +16,59 @@ def create
     render nothing: true, status: 201
   end
 
-  private
+  class KafkaData
+    def initialize(api_data:, controller:)
+      @data = api_data.to_hash.deep_dup
```

Comment:
if you make `KafkaData` inherit from `Hash`, i.e. `KafkaData < Hash` some of this code reduces a bit.

E.g. line 21 becomes

```ruby
merge!(api_data) # may not need deep_dup b/c the values are primitives; also api_data is already a hash
```

The `to_hash` method would go away.  Means in the spec and code you don't have to call `.to_hash` because the `KafkaData` instance is already a hash.

`data[:ip_address] = ` becomes `self[:ip_address] = `

```ruby
source_uri = URI(data[:source_uri])

data[:scheme] = source_uri.scheme
data[:host] = source_uri.host
data[:path] = source_uri.path
data[:query] = source_uri.query ? CGI.parse(source_uri.query) : nil

data.except!(:source_uri)
```

becomes

```ruby
source_uri = URI(delete(:source_uri))

merge!({
  scheme: source_uri.scheme,
  host: source_uri.host,
  path: source_uri.path,
  query: ...
})
```

(even if you don't make `KafkaData < Hash`, the above simplification still works).

This isn't that big of a deal, so you make whatever call you want.

Reply:
Yup, now that I have all this testing in place, this refactor for more idiomatic code is easy!

Reply:
Ok that went pretty easily. One issue rubocop complains that I'm not calling super when subclassing and overriding `initialize`. However, everything "just works" either way. :shrug: